### PR TITLE
Add Jetpack CI infrastructure to target Jetpack tests against staging

### PR DESCRIPTION
**Pull request by dpasque**
Related to #

## Proposed Changes

See discussion here: pd5faL-rf-p2

In short, we want to expand the Jetpack E2E testing we're doing on WPCOM!

Currently, we just use the `TEST_ON_JETPACK` env var flag to switch to an account that has a site hosted on Pressable, but then Jetpack connected to be used in Calypso.

As discussed in the P2, we want to also start running tests against the Jetpack installation on WPCOM, especially in its staging mode during releases.

So, we've made a new env var: `JETPACK_TARGET`. It can be one of three values:
1. `wpcom-production`: this is the jetpack install that any ye-old WPCOM site is using! The baseline, the deafult.
2. `wpcom-staging`: during jetpack releases on WPCOM, a couple of sites have access to the new release in its staging form. We target that staging version on WPCOM here.
3. `not-wpcom`: this represents what we used to target with jetpack testing: a site hosted elsewhere (not on WPCOM) that we've since Jetpack connected and are testing through Calypso. 

There are new CI builds to represent these different options!

We'll be enabling more tests and users for these different cases in a future PR!

## Testing Instructions

- [ ] No TeamCity errors
- [ ] The previous Jetpack tests should continue to target a site hosted on Pressable and pass.

## Pre-merge Checklist

- [x] Has the general commit checklist been followed? (PCYsg-hS-p2)
- [ ] [Have you written new tests](https://wpcalypso.wordpress.com/devdocs/docs/testing/index.md) for your changes?
- [ ] Have you tested the feature in Simple (P9HQHe-k8-p2), Atomic (P9HQHe-jW-p2), and self-hosted Jetpack sites (PCYsg-g6b-p2)?
- [x] Have you checked for TypeScript, React or other console errors?
- [ ] Have you used memoizing on expensive computations? More info in [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors) and [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md)
- [ ] Have we added the "[Status] String Freeze" label as soon as any new strings were ready for translation (p4TIVU-5Jq-p2)?
- [ ] For changes affecting Jetpack: Have we added the "[Status] Needs Privacy Updates" label if this pull request changes what data or activity we track or use (p4TIVU-ajp-p2)?

### CODE REVIEW

**Comment by dpasque** [comment]
Lol, clearly I need to stop using the remote CI servers as my Kotlin compiler 🤦

**Comment by worldomonation** [comment]
> Lol, clearly I need to stop using the remote CI servers as my Kotlin compiler 🤦

Why not? That's what @noahtallen and I both do 🤣

**Comment by dpasque** [comment]
@worldomonation -- it's been a while since I've done one of these PRs, is there any way to see the Kotlin changes for the new build pre-merge? I was in TeamCity looking for this branch, but didn't see them

**Comment by noahtallen** [comment]
> Why not? That's what @noahtallen and I both do 🤣

lmao. On a previous computer, I got it working locally, but I never figured out how to do it again. I should probably take a look again, because it'd be super useful to troubleshoot locally. The main problem is that Gradle or maven or whatever the dependency manager is has to connect to the teamcity server to download the teamcity kotlin library, but that's behind the proxy

**Comment by noahtallen** [comment]
> I was in TeamCity looking for this branch, but didn't see them

Not for new builds, unfortunately :( TeamCity has to know the build exists to be able to trigger it, and it only does that once it's merged to trunk. After that, you can test most changes in the branch

File: .teamcity/_self/projects/WPComTests.kt
Comment (noahtallen) [inline .teamcity/_self/projects/WPComTests.kt]: non-blocking: we could use named parameters here to make it a bit more readable. E.g. `jetpackTarget ="not-wpcom"`. (https://kotlinlang.org/docs/functions.html#named-arguments)

File: .teamcity/_self/projects/WPComTests.kt
Comment (noahtallen) [inline .teamcity/_self/projects/WPComTests.kt]: Potentially better name could be `remote-site`?

File: packages/calypso-e2e/src/env-variables.ts
Comment (noahtallen) [inline packages/calypso-e2e/src/env-variables.ts:118]: Technically, `! supportedValues.includes( output )` is testing that `output` is _not_ the type `JetpackTarget`. (so casting doesn't seem necessary)

**Review by noahtallen — COMMENTED** [review]
This looks pretty solid to me. I think my main concern is just having a lot more builds in our queue. How often would we run these tests? If these will have to run frequently, we'll probably need to get more agents eventually

File: packages/calypso-e2e/src/env-variables.ts
Comment (dpasque) [reply packages/calypso-e2e/src/env-variables.ts:118]: I very much agree! But the TypeScript linter doesn't seem to agree with us ☹️ 

Screen Shot 2023-02-16 at 5 10 02 PM

I figured it was just best to keep the linter happy, even though it feels silly. 🤷  What do you think? I could also cast it to any?

File: .teamcity/_self/projects/WPComTests.kt
Comment (dpasque) [reply .teamcity/_self/projects/WPComTests.kt]: Oh, that's much better lololol

Comment (dpasque) [reply .teamcity/_self/projects/WPComTests.kt]: My faux-enum writing skills never really graduated past the third grade 🤣

**Comment by dpasque** [comment]
Thanks for the review @noahtallen 😀 

>I think my main concern is just having a lot more builds in our queue. How often would we run these tests? If these will have to run frequently, we'll probably need to get more agents eventually

Not very often at all! Each will probably be run only one time each week, literally! They're mostly just used during the Jetpack weekly upgrade. I reckon that shouldn't clog the queue too much?

Also, FYI I broke out several builds with the different parameters mostly to just make it extra simple for the Jetpack folks to be able to run the tests (just go right to the build and click run, as opposed to tweaking a bunch of args). If that's too verbose though, we can collapse these build down.

File: .teamcity/_self/projects/WPComTests.kt
Comment (dpasque) [reply .teamcity/_self/projects/WPComTests.kt]: Fixed in [6acd928](https://github.com/Automattic/wp-calypso/pull/73457/commits/6acd928c6cfc133d195d5578dc07e44821eef125)

File: .teamcity/_self/projects/WPComTests.kt
Comment (dpasque) [reply .teamcity/_self/projects/WPComTests.kt]: Fixed in [6acd928](https://github.com/Automattic/wp-calypso/pull/73457/commits/6acd928c6cfc133d195d5578dc07e44821eef125)

**Comment by worldomonation** [comment]
> How often would we run these tests?

To add a bit more context, this is for pd5faL-qu-p2 and this particular build will be triggered only once a week or so, manually.

File: packages/calypso-e2e/src/env-variables.ts
Comment (noahtallen) [reply packages/calypso-e2e/src/env-variables.ts:118]: I think we might as well let `supportedValues` be a plain string array unless having the type info is otherwise useful (for that variable)

**Comment by noahtallen** [comment]
Cool!

File: test/e2e/specs/appearance/theme__details-preview-activate.ts
Comment (worldomonation) [inline test/e2e/specs/appearance/theme__details-preview-activate.ts:26]: **thought:non-blocking** - it occurred to me that we rely on this a lot, but do we actually have unit tests within `calypso-e2e` to verify this behavior?

**Review by worldomonation — APPROVED** [review]

File: packages/calypso-e2e/src/env-variables.ts
Comment (dpasque) [reply packages/calypso-e2e/src/env-variables.ts:118]: > unless having the type info is otherwise useful (for that variable)

Ah, yeah, the main reason I type it as such is that I was hoping for transpile-time safety for that variable in case we update the options in the future. Otherwise, we could end up in a scenario where we change the typing, but that array sneaks through, and we accept (at runtime) options that aren't actually valid. 

The cast definitely looks weird, but my personal preference is currently to take the type safety at the expense of the weird-looking cast! But, definitely not married to that, so I'd be glad to swing the tradeoff back the other way!

Comment (dpasque) [reply packages/calypso-e2e/src/env-variables.ts:118]: Also, I'm going to merge this to unblock a few things, but I'm glad to go back and update this too if we decide that is better! Just let me know 😄

**Pull request merged**
